jwst/lib/basic_utils.py

```diff
@@ -1,9 +1,30 @@
 """General utility objects"""
 
+from stdatamodels.jwst import datamodels
 from stdatamodels.jwst.datamodels import dqflags
 import numpy as np
 
 
+def use_datamodel(input):
```

Comment:
What about updating `JwstStep._datamodels_open` and using it (or `Step.open_model`) instead of adding this utility function? That would have the added benefit of applying a similar fix (avoiding unnecessary model clones triggered by `datamodels.open(Model)`) to the calls that stpipe makes to open models to determine references.

Reply:
One other benefit to updating and using `Step.open_model` would be to allow the opening code to check if the step is part of a pipeline (where `Step.parent` will return the pipeline). This might be too "fancy" (and harder to make sense of) but the general idea would be to replace the following (found in many of the steps)
```
with datamodels.open(input) as input:
```
with
```
with self.open_model(input) as input:
```
And implement `_datamodels_open` in a way that:
- when `Step.parent` is none, mimic the old behavior (create a clone when a model is provided)
- when `Step.parent` is not None only open the model if a string (or path) is provided
- optionally provide a step class (like `RampModel`) to use when opening the model (or to create a clone when the input is not an instance of the class)
- if a model was not opened (or a clone made) return the model with a `nullcontext` so that the input is not closed when the step finishes

Dropping the old behavior (cloning the input) is worth considering because clones share data arrays (which are often updated in steps).

Reply:
I implemented your suggestions, thank you!

Reply:
Thanks! Looking at `_datamodels_open` for this PR:
https://github.com/spacetelescope/jwst/pull/8588/files#diff-9a8d6de889fab1db67280357ed291190cbcf4fa320977c44c922db8cc78976a6R28-R33
the items described in the bullet points above are missing.

> - when Step.parent is none, mimic the old behavior (create a clone when a model is provided)
> - when Step.parent is not None only open the model if a string (or path) is provided

I don't see any check to `Step.parent`. This means every step now. modifies the input when provided a `DataModel` (even when run outside the pipeline).

> - optionally provide a step class (like RampModel) to use when opening the model (or to create a clone when the input is not an instance of the class)

I don't see any option for a model class but I think given the limited use of this behavior (forcing a model type) it might be cleaner to just have this in detector 1.

> - if a model was not opened (or a clone made) return the model with a nullcontext so that the input is not closed when the step finishes

This one is quite important as most steps use a `with` statement. If a `nullcontext` isn't used the steps will now close the input model when they finish. Since this is the same as the output model the next step in the pipeline will receive a closed model. This can cause things to crash for different model/step combinations as the full contents of a model might not be read when the file is opened.

Reply:
this has been implemented

Reply:
the function that does this is copy_datamodel(input, step_parent), it checks if self.parent is None and creates a copy, otherwise it returns the input

Reply:
in other words, the procedure now is to use use_datamodel(input, model_class) and/or make sure the model is the desired class, then make a copy or not based on whether the step is part of a pipeline or not